File: backend/app/schemas/vehicles.py
"""Pydantic schemas for vehicle endpoints (Task B1)."""

from datetime import date, datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

FUEL_TYPES = {"petrol", "diesel", "electric", "hybrid", "other"}


class VehicleBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    make: str | None = None
    model: str | None = None
    year: int | None = None
    registration_number: str | None = Field(default=None, alias="registrationNumber")
    vin: str | None = None
    purchase_date: date | None = Field(default=None, alias="purchaseDate")
    purchase_price_cents: int | None = Field(default=None, alias="purchasePriceCents")
    currency: str = "USD"
    current_mileage: int | None = Field(default=None, alias="currentMileage")
    vehicle_type: str | None = Field(default=None, alias="vehicleType")
    fuel_type: str | None = Field(default=None, alias="fuelType")
    default_fuel_variant: str | None = Field(default=None, alias="defaultFuelVariant")
    distance_unit: str | None = Field(default=None, alias="distanceUnit")
    photo_url: str | None = Field(default=None, alias="photoUrl")
    photo_public_id: str | None = Field(default=None, alias="photoPublicId")

    @field_validator("fuel_type")
    @classmethod
    def validate_fuel_type(cls, v: str | None) -> str | None:
        if v is not None and v not in FUEL_TYPES:
            raise ValueError(f"fuelType must be one of {sorted(FUEL_TYPES)}")
        return v

    @field_validator("distance_unit")
    @classmethod
    def validate_distance_unit(cls, v: str | None) -> str | None:
        if v is not None and v not in {"km", "mi"}:
            raise ValueError('distanceUnit must be "km", "mi", or null')
        return v


class VehicleCreate(VehicleBase):
    make: str
    model: str


class VehicleUpdate(VehicleBase):
    pass


class VehicleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    make: str
    model: str
    year: int | None = None
    registration_number: Annotated[
        str | None, Field(default=None, serialization_alias="registrationNumber")
    ]
    vin: str | None = None
    purchase_date: Annotated[date | None, Field(default=None, serialization_alias="purchaseDate")]
    purchase_price_cents: Annotated[
        int | None, Field(default=None, serialization_alias="purchasePriceCents")
    ]
    currency: str
    current_mileage: Annotated[
        int | None, Field(default=None, serialization_alias="currentMileage")
    ]
    vehicle_type: Annotated[str | None, Field(default=None, serialization_alias="vehicleType")]
    fuel_type: Annotated[str | None, Field(default=None, serialization_alias="fuelType")]
    default_fuel_variant: Annotated[
        str | None, Field(default=None, serialization_alias="defaultFuelVariant")
    ]
    distance_unit: Annotated[str | None, Field(default=None, serialization_alias="distanceUnit")]
    photo_url: Annotated[str | None, Field(default=None, serialization_alias="photoUrl")]
    photo_public_id: Annotated[str | None, Field(default=None, serialization_alias="photoPublicId")]
    created_at: Annotated[datetime, Field(serialization_alias="createdAt")]
    updated_at: Annotated[datetime, Field(serialization_alias="updatedAt")]

File: web/src/server/schemas/vehicles.ts
import { z } from 'zod';

const FUEL_TYPES = ['petrol', 'diesel', 'electric', 'hybrid', 'other'] as const;

/** Derived document-expiry health for a vehicle (Doc 3). */
export const docsStatusSchema = z.object({
  state: z.enum(['none', 'valid', 'needs_action']),
  needsActionCount: z.number().int().nonnegative(),
});

export type DocsStatus = z.infer<typeof docsStatusSchema>;

const vehicleFields = {
  make: z.string().min(1),
  model: z.string().min(1),
  year: z.number().int().min(1900).max(2100).nullable().optional(),
  registrationNumber: z.string().nullable().optional(),
  vin: z.string().nullable().optional(),
  purchaseDate: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, 'purchaseDate must be YYYY-MM-DD')
    .nullable()
    .optional(),
  purchasePriceCents: z.number().int().nullable().optional(),
  currency: z
    .string()
    .regex(/^[A-Za-z]{3}$/, 'currency must be a 3-letter code')
    .optional(),
  currentMileage: z.number().int().nullable().optional(),
  vehicleType: z.string().nullable().optional(),
  fuelType: z.enum(FUEL_TYPES).nullable().optional(),
  distanceUnit: z.enum(['km', 'mi']).nullable().optional(),
  photoUrl: z.string().nullable().optional(),
  photoPublicId: z.string().nullable().optional(),
};

/** POST /api/v1/vehicles — make + model required. */
export const createVehicleSchema = z.object(vehicleFields);

export type CreateVehicle = z.infer<typeof createVehicleSchema>;

/** PATCH /api/v1/vehicles/{id} — all fields optional. */
export const updateVehicleSchema = z.object({
  make: z.string().min(1).optional(),
  model: z.string().min(1).optional(),
  year: z.number().int().min(1900).max(2100).nullable().optional(),
  registrationNumber: z.string().nullable().optional(),
  vin: z.string().nullable().optional(),
  purchaseDate: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, 'purchaseDate must be YYYY-MM-DD')
    .nullable()
    .optional(),
  purchasePriceCents: z.number().int().nullable().optional(),
  currency: z
    .string()
    .regex(/^[A-Za-z]{3}$/, 'currency must be a 3-letter code')
    .optional(),
  currentMileage: z.number().int().nullable().optional(),
  vehicleType: z.string().nullable().optional(),
  fuelType: z.enum(FUEL_TYPES).nullable().optional(),
  distanceUnit: z.enum(['km', 'mi']).nullable().optional(),
  photoUrl: z.string().nullable().optional(),
  photoPublicId: z.string().nullable().optional(),
});

export type UpdateVehicle = z.infer<typeof updateVehicleSchema>;

/** Wire shape for Vehicle responses (Doc 3). */
export const vehicleResponseSchema = z.object({
  id: z.string().uuid(),
  make: z.string(),
  model: z.string(),
  year: z.number().nullable(),
  registrationNumber: z.string().nullable(),
  vin: z.string().nullable(),
  purchaseDate: z.string().nullable(),
  purchasePriceCents: z.number().nullable(),
  currency: z.string(),
  currentMileage: z.number().nullable(),
  vehicleType: z.string().nullable(),
  fuelType: z.string().nullable(),
  distanceUnit: z.enum(['km', 'mi']).nullable(),
  photoUrl: z.string().nullable(),
  photoPublicId: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  docsStatus: docsStatusSchema,
});

export type VehicleResponse = z.infer<typeof vehicleResponseSchema>;
